xarray/core/accessor_str.py

```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        ValueError : `dim` is `None` and there is more than one capture group.
+        ValueError : `case` is set when `pat` is a compiled regular expression.
+        KeyError : The given dimension is already present in the DataArray.
+
+        Examples
+        --------
+        Create a string array
+
+        >>> value = xr.DataArray(
+        ...     [
+        ...         [
+        ...             "a_Xy_0",
+        ...             "ab_xY_10-bab_Xy_110-baab_Xy_1100",
+        ...             "abc_Xy_01-cbc_Xy_2210",
+        ...         ],
+        ...         [
+        ...             "abcd_Xy_-dcd_Xy_33210-dccd_Xy_332210",
+        ...             "",
+        ...             "abcdef_Xy_101-fef_Xy_5543210",
+        ...         ],
+        ...     ],
+        ...     dims=["X", "Y"],
+        ... )
+
+        Extract matches
+
+        >>> value.str.extract(r"(\\w+)_Xy_(\\d*)", dim="XX")
+        <xarray.DataArray (X: 2, Y: 3, XX: 2)>
+        array([[['a', '0'],
+                ['bab', '110'],
+                ['abc', '01']],
+        <BLANKLINE>
+               [['abcd', ''],
+                ['', ''],
+                ['abcdef', '101']]], dtype='<U6')
+        Dimensions without coordinates: X, Y, XX
+
+        See Also
+        --------
+        DataArray.str.extractall
+        DataArray.str.findall
+        re.compile
+        re.search
+        pandas.Series.str.extract
+        """
+        pat = self._re_compile(pat, flags, case)
+
+        if pat.groups == 0:
+            raise ValueError("No capture groups found in pattern.")
+
+        if dim is None and pat.groups != 1:
+            raise ValueError("dim must be specified if more than one capture group.")
+
+        if dim is not None and dim in self._obj.dims:
+            raise KeyError(f"Dimension {dim} already present in DataArray.")
+
+        def _get_res_single(val, pat=pat):
+            match = pat.search(val)
+            if match is None:
+                return ""
+            res = match.group(1)
+            if res is None:
+                res = ""
+            return res
+
+        def _get_res_multi(val, pat=pat):
+            match = pat.search(val)
+            if match is None:
+                return np.array([""], val.dtype)
+            match = match.groups()
+            match = [grp if grp is not None else "" for grp in match]
+            return np.array(match, val.dtype)
+
+        if dim is None:
+            return self._apply(_get_res_single)
+        else:
+            # dtype MUST be object or strings can be truncated
+            # See: https://github.com/numpy/numpy/issues/8352
+            return self._apply(
+                _get_res_multi,
+                dtype=np.object_,
+                output_core_dims=[[dim]],
+                output_sizes={dim: pat.groups},
+            ).astype(self._obj.dtype.kind)
+
+    def extractall(
+        self,
+        pat: Any,
+        group_dim: Hashable,
+        match_dim: Hashable,
```

Comment:
```suggestion
        group_dim: Hashable = "group",
        match_dim: Hashable = "match",
```
I'd vote for giving sensible defaults here.